src/Magnum/DartIntegration/ConvertShapeNode.cpp

```diff
@@ -26,167 +26,306 @@
+        Containers::Array<Trade::PhongMaterialData> materials(Containers::NoInit, meshesCount);
+
+        UnsignedInt j = 0;
+        for(UnsignedInt i = 0; i < importer->object3DCount(); i++) {
+            auto meshData3D = dynamic_cast<Trade::MeshObjectData3D*>(importer->object3D(i).release());
+            if(meshData3D) {
+                Containers::Optional<Trade::MeshData3D> meshData = importer->mesh3D(meshData3D->instance());
+                if(!meshData)
+                    return Containers::NullOpt;
+
+                if(getMaterial) {
+                    auto colorMode = meshShape->getColorMode();
+                    /* only get materials from mesh if the appropriate color mode */
+                    if(importer->materialCount() && getMaterial) {
+                        if(colorMode == dart::dynamics::MeshShape::ColorMode::MATERIAL_COLOR) {
+                            auto matPtr = importer->material(meshData3D->material());
+                            new(&materials[j]) Trade::PhongMaterialData{std::move(*static_cast<Trade::PhongMaterialData*>(matPtr.get()))};
+                        }
+                        else if(colorMode == dart::dynamics::MeshShape::ColorMode::COLOR_INDEX) {
+                            /* @todo: check if index is within bounds */
+                            /* get diffuse color from Mesh color */
+                            Color4 meshColor = meshData->colors(0)[meshShape->getColorIndex()];
+                            new(&materials[j]) Trade::PhongMaterialData{Trade::PhongMaterialData::Flags{}, 80.f};
+                            materials[j].diffuseColor() = Color3(meshColor[0], meshColor[1], meshColor[2]);
+                            /* default colors for ambient (black) and specular (white) */
+                            materials[j].ambientColor() = Vector3{0.f, 0.f, 0.f};
+                            materials[j].specularColor() = Vector3{1.f, 1.f, 1.f};
+                        }
+                        else if (colorMode == dart::dynamics::MeshShape::ColorMode::SHAPE_COLOR) {
+                            new(&materials[j]) Trade::PhongMaterialData{std::move(nodeMaterial)};
+                        }
+                    }
+                }
+                /* fallback material to by-pass seg-faults */
+                else
+                    new(&materials[j]) Trade::PhongMaterialData{Trade::PhongMaterialData::Flags{}, 80.f};
+
+                if(getMesh) {
+                    new(&meshes[j]) Trade::MeshData3D{std::move(*meshData)};
+                }
+                /* fallback mesh data to by-pass seg-faults */
+                else
+                    new(&meshes[j]) Trade::MeshData3D{MeshPrimitive::Triangles, std::vector<UnsignedInt>(), std::vector<std::vector<Vector3>>(1, std::vector<Vector3>()), std::vector<std::vector<Vector3>>(), std::vector<std::vector<Vector2>>(), std::vector<std::vector<Color4>>()};
```

Comment:
> One more argument for Optional: if someone is using the ShapeData directly, they currently wouldn't have a intuitive way to know if a particular mesh is valid or not. With Optional this would be obvious.

If a mesh is asked to be loaded and cannot be loaded, we return `Containers::NullOpt`. Otherwise the size of the `ShapeData::meshes` gives us the number of loaded meshes (`Trade::MeshData3D`) and all of them are always valid..

> As we discussed before -- wouldn't Containers::Array<Containers::Optional<Trade::MeshData3D> (and similar for materials) be a better / safer solution here, to avoid such suspiciously-looking workarounds?

To this I mostly agree, so I am also inclining to change things to `Containers::Optional`..